### ISSUE TITLE
Check if opaque

### ISSUE THREAD

**Issue by gsakkis** [openslide/openslide#424]
### Issue details

OpenSlide always returns an [RGBA Image](https://openslide.org/api/python/#openslide.OpenSlide.read_region), even if the image happens to be opaque. Is there a way to determine if the image is opaque before/without actually reading it?

**Comment by ryanlandvater** [comment]
OpenSlide always writes 32 bit per pixel. In the event of an opaque image, it fills the alpha byte with 0xFF. See openslide-decode-jpeg.c line 288 for an example of that. The query would not help unfortunately. Most rendering pipelines expect 32 bits per pixel so unless you have / wrote a rendering pipeline with a 24 bit stride cutting out the extra byte would probably not help much.

**Comment by gsakkis** [comment]
The (implied) feature request here is not to allow writing 24 bits/pixel, it's a `bool is_opaque()` (or somesuch) method/function I would be able to call before calling `read_region`.

**Comment by bgilbert** [comment]
As you've worded it, it sounds like you're looking to know whether every pixel in a region is 100% opaque.  That isn't really a meaningful query.  Some pixels might be partially transparent as a result of compositing, but still have useful contents.

If you want to know whether an arbitrary region is fully transparent (i.e. has no interesting pixels), there isn't currently a way to make that query, but you can use the [bounds properties](https://openslide.org/docs/properties/) to get a bounding box around the portion of the slide that does contain pixel data.

**Comment by gsakkis** [comment]
I'm probably still not wording it correctly so let me try again with PIL using [CMU-1-Small-Region.svs](https://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/CMU-1-Small-Region.svs) as sample file:

```py
In [1]: from PIL import Image

In [2]: Image.open("CMU-1-Small-Region.svs")
Out[2]: <PIL.TiffImagePlugin.TiffImageFile image mode=RGB size=2220x2967>

In [3]: Image.open("CMU-1-Small-Region.svs").mode
Out[3]: 'RGB'

In [4]: # convert it to RGBA with tiff2rgba

In [5]: !tiff2rgba CMU-1-Small-Region.svs new-CMU-1-Small-Region.svs

In [6]: Image.open("new-CMU-1-Small-Region.svs")
Out[6]: <PIL.TiffImagePlugin.TiffImageFile image mode=RGBA size=2220x2967>

In [7]: Image.open("new-CMU-1-Small-Region.svs").mode
Out[7]: 'RGBA'
```

So PIL can somehow figure out that the original file's mode is RGB and the converted one's is RGBA. Can OpenSlide do the same?

**Comment by bgilbert** [comment]
Ah, I see.  The fact that OpenSlide returns an ARGB image (RGBA in OpenSlide Python) has nothing to do with the pixel format of the underlying file.  OpenSlide needs the alpha channel to record the extent to which image data is present for a pixel, since some of the supported file formats allow image tiles to be missing.  So, a pixel that has image data would be 100% opaque, a pixel in an area the scanner didn't capture would be 0% opaque, and a pixel downsampled from one of the former and three of the latter would be 25% opaque.  Because of subpixel positioning, arbitrary opacity values are possible, though most pixels of interest will be 100% opaque or close to it.

**Comment by gsakkis** [comment]
> The fact that OpenSlide returns an ARGB image (RGBA in OpenSlide Python) has nothing to do with the pixel format of the underlying file. 

Right, and the question I've been trying to get an answer for (unsuccessfully so far) is: 

**Does OpenSlide expose programmatically the pixel format of the underlying file?**

I'd gladly use PIL (or even parse the `tiffinfo` output) for this but I'm guessing they don't support all the formats that OpenSlide does, hence this issue.

**Comment by bgilbert** [comment]
It does not.  OpenSlide is intended to be an abstraction layer, as much as possible, so that you don't need to deal with the technical details of individual slide formats.

**Comment by gsakkis** [comment]
Thank you. I was hoping it would be somehow exposed, just like a bunch of other metadata:
```
In [3]: dict(openslide.OpenSlide("CMU-1-Small-Region.svs").properties)
Out[3]: 
{'aperio.AppMag': '20',
 'aperio.Date': '12/29/09',
 'aperio.Filename': 'CMU-1',
 'aperio.Filtered': '5',
 'aperio.Focus Offset': '0.000000',
 'aperio.ImageID': '1004486',
 'aperio.Left': '25.691574',
 'aperio.LineAreaXOffset': '0.019265',
 'aperio.LineAreaYOffset': '-0.000313',
 'aperio.LineCameraSkew': '-0.000424',
 'aperio.MPP': '0.4990',
 'aperio.OriginalHeight': '32914',
 'aperio.OriginalWidth': '46000',
 'aperio.Originalheight': '33014',
 'aperio.Parmset': 'USM Filter',
 'aperio.ScanScope ID': 'CPAPERIOCS',
 'aperio.StripeWidth': '2040',
 'aperio.Time': '09:59:15',
 'aperio.Top': '23.449873',
 'aperio.User': 'b414003d-95c6-48b0-9369-8010ed517ba7',
 'openslide.comment': 'Aperio Image Library v11.2.1 \r\n46000x32914 [42673,5576 2220x2967] (240x240) JPEG/RGB Q=30;Aperio Image Library v10.0.51\r\n46920x33014 [0,100 46000x32914] (256x256) JPEG/RGB Q=30|AppMag = 20|StripeWidth = 2040|ScanScope ID = CPAPERIOCS|Filename = CMU-1|Date = 12/29/09|Time = 09:59:15|User = b414003d-95c6-48b0-9369-8010ed517ba7|Parmset = USM Filter|MPP = 0.4990|Left = 25.691574|Top = 23.449873|LineCameraSkew = -0.000424|LineAreaXOffset = 0.019265|LineAreaYOffset = -0.000313|Focus Offset = 0.000000|ImageID = 1004486|OriginalWidth = 46920|Originalheight = 33014|Filtered = 5|OriginalWidth = 46000|OriginalHeight = 32914',
 'openslide.level-count': '1',
 'openslide.level[0].downsample': '1',
 'openslide.level[0].height': '2967',
 'openslide.level[0].tile-height': '240',
 'openslide.level[0].tile-width': '240',
 'openslide.level[0].width': '2220',
 'openslide.mpp-x': '0.499',
 'openslide.mpp-y': '0.499',
 'openslide.objective-power': '20',
 'openslide.quickhash-1': '6335ea0e6cc54c2cba64bb265d3c713a50cd84484924e3a9c109558c13521d5c',
 'openslide.vendor': 'aperio',
 'tiff.ImageDescription': 'Aperio Image Library v11.2.1 \r\n46000x32914 [42673,5576 2220x2967] (240x240) JPEG/RGB Q=30;Aperio Image Library v10.0.51\r\n46920x33014 [0,100 46000x32914] (256x256) JPEG/RGB Q=30|AppMag = 20|StripeWidth = 2040|ScanScope ID = CPAPERIOCS|Filename = CMU-1|Date = 12/29/09|Time = 09:59:15|User = b414003d-95c6-48b0-9369-8010ed517ba7|Parmset = USM Filter|MPP = 0.4990|Left = 25.691574|Top = 23.449873|LineCameraSkew = -0.000424|LineAreaXOffset = 0.019265|LineAreaYOffset = -0.000313|Focus Offset = 0.000000|ImageID = 1004486|OriginalWidth = 46920|Originalheight = 33014|Filtered = 5|OriginalWidth = 46000|OriginalHeight = 32914',
 'tiff.ResolutionUnit': 'inch'}
```

Could a new (say) `openslide.samples_per_pixel` property be added?

**Comment by bgilbert** [comment]
Taking a step back for a moment, could you explain your use case?  What would you do with this information if it was available?  I'm not aware of any whole-slide image format that stores transparency data (rather than just the set of tiles that are present or absent), and if one exists, I wouldn't think it'd affect how you'd process OpenSlide's output.

**Comment by gsakkis** [comment]
The use case involves representing the image data as a multidimensional array , both in memory (Numpy) and on disk (TileDB). Currently for OpenSlide this needs to be a 3D `height x width x 4` array. Knowing that the image is opaque would allow representing it as `height x width x 3`, yielding 25% memory savings for uncompressed data (admittedly the gain would be negligible for compressed data).

**Comment by bgilbert** [comment]
Okay, makes sense.  As I said, though, the existence of an alpha channel _in the underlying tiles_ isn't what controls whether you need to retain the alpha.  OpenSlide will produce a meaningful alpha channel for any format that allows missing tiles or tile placement outside of a regular grid (which is many of them), even if the individual tiles are opaque.

I'd recommend always treating the alpha as meaningful.  You might choose to avoid storing it, though, by compositing the pixel data onto a white background (as many viewer applications do).

**Comment by gsakkis** [comment]
> OpenSlide will produce a meaningful alpha channel for any format that allows missing tiles or tile placement outside of a regular grid (which is many of them), even if the individual tiles are opaque.

Ah interesting, based on my (limited) tests and @ryanlandvater's reply I thought alpha would always be filled with 0xFF.

I will follow your recommendation and preserve alpha, feel free to close this issue. Thanks!

**Comment by bgilbert** [comment]
Okay, sounds good!
